From f3cfb3abcb6513bbaff71a5b91e15b06a5daa15b Mon Sep 17 00:00:00 2001
From: steven_fann <steven_fann@realtek.com>
Date: Wed, 20 Jan 2021 14:43:00 +0800
Subject: [PATCH] Do not destroy adb socket when it is running over TCP.

Bug: b/161861298
Test: atest android.net.cts.Ikev2VpnTest#testStartStopVpnProfileV4

Change-Id: I595c61bf0a08882a62aebec86db8ab2965fe6a57
Tracked-On: OAM-95330
Signed-off-by: Aiswarya Cyriac<aiswarya.cyriac@inte.com>
---
 server/SockDiag.cpp | 13 ++++++++++++-
 1 file changed, 12 insertions(+), 1 deletion(-)

diff --git a/server/SockDiag.cpp b/server/SockDiag.cpp
index 44bda3b0..e0b6b4b7 100644
--- a/server/SockDiag.cpp
+++ b/server/SockDiag.cpp
@@ -31,6 +31,7 @@
 
 #include <cinttypes>
 
+#include <android-base/properties.h>
 #include <android-base/strings.h>
 #include <log/log.h>
 #include <netdutils/InternetAddresses.h>
@@ -52,6 +53,15 @@ using netdutils::Stopwatch;
 namespace net {
 namespace {
 
+int getAdbPort() {
+    return android::base::GetIntProperty("service.adb.tcp.port", 0);
+}
+
+bool isAdbSocket(const inet_diag_msg *msg, int adbPort) {
+    return adbPort > 0 && msg->id.idiag_sport == htons(adbPort) &&
+        (msg->idiag_uid == AID_ROOT || msg->idiag_uid == AID_SHELL);
+}
+
 int checkError(int fd) {
     struct {
         nlmsghdr h;
@@ -414,7 +424,8 @@ int SockDiag::destroySockets(const UidRanges& uidRanges, const std::set<uid_t>&
         return msg != nullptr &&
                uidRanges.hasUid(msg->idiag_uid) &&
                skipUids.find(msg->idiag_uid) == skipUids.end() &&
-               !(excludeLoopback && isLoopbackSocket(msg));
+               !(excludeLoopback && isLoopbackSocket(msg)) &&
+               !isAdbSocket(msg, getAdbPort());
     };
 
     iovec iov[] = {
-- 
2.17.1

